sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala

```diff
@@ -60,18 +58,14 @@ object ExtractValue {
         GetArrayStructFields(child, fields(ordinal).copy(name = fieldName),
           ordinal, fields.length, containsNull)
 
-      case (_: ArrayType, _) if extraction.dataType.isInstanceOf[IntegralType] =>
-        GetArrayItem(child, extraction)
+      case (_: ArrayType, _) => GetArrayItem(child, extraction)
```

Comment:
a small but unrelated bug: the `extraction` expression may not be literal and maybe unresolved, so we can't call `dataType` of `extraction` here, should leave it to the type checking interface.
I also improved the test `DataFrameSuite.SPARK-7133: Implement struct, array, and map field accessor`, to test non-literal `extraction` expression on both `GetArrayItem` and `GetMapValue`.